working/extension_structs/overview.md

```diff
@@ -0,0 +1,870 @@
+discussion.  If there is buy-in on these ideas, we may choose to incorporate
+them into an existing proposal, or add a new feature specification.
+
+
+## Design principles
+
+We aim to minimize the differences between structs and classes.  As much as
+possible, structs should behave as restrictions of classes, and extension
+structs should behave as further restrictions of structs.  We specifically aim
+to avoid as much as possible having different behaviors for the same concept
+(e.g. differences in scoping).
+
+We also aim to minimize the amount of new syntax required, and to maximize the
+amount of new functionality that we can provide relative to the syntactic real
+estate consumed, and the new cognitive load imposed on users.
+
+## Roadmap
+
+The first section describes the full struct feature.  The second section
+describes the restriction of the full struct to support static wrappers.
+
+## Structs
+
+This section describes how structs look and behave, largely by example.  In many
+cases, alternatives or extensions to the core proposal are described in line.
+Larger extensions are postponed to a later section.
+
+## Structs by example
+
+Structs allow you to define simple classes containing immutable data very
+compactly.  For example, we might model a component vector as follows:
+
+```
+struct Component(int x, int y, int z, int w = 1);
+```
+
+This short definition essentially defines a class `Component`, with four fields,
+and provides a number of convenient methods.
+
+```dart
+void test() {
+  // Call the default constructor, using the default value for w
+  var c1 = Component(0, 0, 0);
+  // Call the default constructor, explicitly passing w
+  var c2 = Component(1, 1, 1, 1);
+
+  assert(c1 == c2); // Equality is defined to be structural
+  assert(c1.hashCode == c2.hashCode); // With correspondind hashCode
+
+  print(c1.debugToString()); // Prints "Component(0, 0, 0, 1)"
+
+  print(c1.toString()); // Prints "struct"
+
+  // Generated copy method
+  var c2 = c1.copyWith(x : 2);
+  assert(c2.x == 2);
+}
+```
+
+Structs may explicitly define constructors, static members, and normal class
+members (except fields), and may implement interfaces.  All fields must be
+declared in the header (the "primary constructor"), and are implicitly final
+(making all structs shallowly immutable).  Fields may be marked as private as
+usual by naming them with a leading `_` in their name.
+
+```dart
+struct Component(int _x, int _y, int _z, int _w = 1)
+  implements Comparable<Component> {
+
+  double get x => _x/_w;
+  double get y => _x/_w;
+  double get z => _z/_w;
+
+  int compareTo(Component other) => throw "TODO: Unimplemented";
+
+  Component.zero() : _x = 0, _y = 0, _z = 0;
+}
+```
+
+With the addition of pattern matching and switches, structs will support closed
+families (algebraic datatypes).
+
+```dart
+abstract struct Operand;
+
+struct ConstantOperand(Value c) extends Operand;
+
+struct IdentifierOperand(Identifier i) extends Operand;
```

Comment:
Yeah, this would be really nice.

Reply:
Algebraic types tend to grow to be a discriminated union over many elements. dart2js's SSA HInstruction type has O(100) subclasses in a rich hierarchy. 'pattern matching' is done via a visitor pattern that has base class visitors that delegate to a 'visit' method for the abstract superclass. This is all tedious and error-prone.

This kind of hierarchy is often thousands of lines of code with each element of the hierarchy implementing dozens of methods.

It would be great if the class hierarchy + visitor patterns could be replaced by a hierarchical algebraic (but partially mutable) types.

What would be most disappointing is if a simple hierarchy could be written using something like the proposed syntax above but as the code-base evolves at some point it needs to be completely rewritten in the hierarchy+visitor pattern.

Can multilevel algebraic types with mutable fields be supported?

```dart
class Operand switch {
  bool get isSimple;  // abstract member that must be implemented by each 'case'
  case ConstantOperand switch {
    bool get isSimple => true;
    method() { ... }  // methodapplicable to all ConstantOperands
    case StringConstantOperand(String stringValue) {
      List<int> get utf8Bytes => ...;
    }
    case IntConstantOperand(int intValue) {
    }
    ...
  }
  ...
}
```

Reply:
> Can multilevel algebraic types with mutable fields be supported?

I think there are two separate pieces of your question.  I think the patterns/sealed family proposal from @munificent would completely support your use case: you get hierarchies of normal classes, with exhaustiveness checking.  The one restriction is that they need to be in the same library.

There's a separate question around a nice short (and possibly hierarchical) syntax for specifying a sealed family.  I think @munificent has some ideas around that as well, along the lines of what you sketch out above.